// better-logging/dist/lib/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveConfig = exports.DefaultConfig = exports.Config = void 0;
const typescript_helpers_1 = require("@olian/typescript-helpers");
const messageConstructionStrategy_1 = require("./enums/messageConstructionStrategy");
const dark_1 = require("../themes/dark");
class Config extends typescript_helpers_1.Record {
}
exports.Config = Config;
exports.DefaultConfig = new Config({
    messageConstructionStrategy: messageConstructionStrategy_1.MessageConstructionStrategy.ALL,
    format: ctx => `${ctx.time24} ${ctx.type} ${ctx.msg}`,
    saveToFile: null,
    color: dark_1.theme,
    logLevels: {
        debug: 4,
        log: 3,
        info: 2,
        line: 1,
        warn: 1,
        error: 0
    },
});
exports.resolveConfig = (config) => new Config({
    messageConstructionStrategy: typescript_helpers_1.useValueOrFallback(config, 'messageConstructionStrategy', exports.DefaultConfig.messageConstructionStrategy),
    format: typescript_helpers_1.useValueOrFallback(config, 'format', exports.DefaultConfig.format),
    saveToFile: typescript_helpers_1.useValueOrFallback(config, 'saveToFile', null),
    color: !config.color ? exports.DefaultConfig.color : {
        base: typescript_helpers_1.useValueOrFallback(config.color, 'base', exports.DefaultConfig.color.base),
        type: !config.color.type ? exports.DefaultConfig.color.type : {
            debug: typescript_helpers_1.useValueOrFallback(config.color.type, 'debug', exports.DefaultConfig.color.type.debug),
            error: typescript_helpers_1.useValueOrFallback(config.color.type, 'error', exports.DefaultConfig.color.type.error),
            info: typescript_helpers_1.useValueOrFallback(config.color.type, 'info', exports.DefaultConfig.color.type.info),
            log: typescript_helpers_1.useValueOrFallback(config.color.type, 'log', exports.DefaultConfig.color.type.log),
            warn: typescript_helpers_1.useValueOrFallback(config.color.type, 'warn', exports.DefaultConfig.color.type.warn),
        }
    },
    logLevels: !config.logLevels ? exports.DefaultConfig.logLevels : {
        debug: typescript_helpers_1.useValueOrFallback(config.logLevels, 'debug', exports.DefaultConfig.logLevels.debug),
        error: typescript_helpers_1.useValueOrFallback(config.logLevels, 'error', exports.DefaultConfig.logLevels.error),
        info: typescript_helpers_1.useValueOrFallback(config.logLevels, 'info', exports.DefaultConfig.logLevels.info),
        log: typescript_helpers_1.useValueOrFallback(config.logLevels, 'log', exports.DefaultConfig.logLevels.log),
        warn: typescript_helpers_1.useValueOrFallback(config.logLevels, 'warn', exports.DefaultConfig.logLevels.warn),
        line: typescript_helpers_1.useValueOrFallback(config.logLevels, 'line', exports.DefaultConfig.logLevels.line),
    }
});
